qcodes/instrument/visa.py

```diff
@@ -74,51 +69,66 @@ def __init__(
                                                vals.Enum(None)))
 
         try:
-            self.set_address(address)
+            visa_handle, visabackend = self._open_resource(address, visalib)
         except Exception as e:
             self.visa_log.exception(f"Could not connect at {address}")
             self.close()
             raise e
 
+        self.visabackend: str = visabackend
+        self.visa_handle: pyvisa.resources.MessageBasedResource = visa_handle
+        """
+        The VISA resource used by this instrument.
+        """
+        self.visalib: Optional[str] = visalib
+        self._address = address
+
         if device_clear:
             self.device_clear()
 
         self.set_terminator(terminator)
         self.timeout.set(timeout)
 
-    def set_address(self, address: str) -> None:
-        """
-        Set the address for this instrument.
-
-        Args:
-            address: The visa resource name to use to connect. The address
-                should be the actual address and just that. If you wish to
-                change the backend for VISA, use the self.visalib attribute
-                (and then call this function).
-        """
+    def _open_resource(
+        self, address: str, visalib: Optional[str]
+    ) -> Tuple[pyvisa.resources.MessageBasedResource, str]:
 
         # in case we're changing the address - close the old handle first
-        if getattr(self, 'visa_handle', None):
+        if getattr(self, "visa_handle", None):
             self.visa_handle.close()
 
-        if self.visalib:
+        if visalib is not None:
             self.visa_log.info(
-                f"Opening PyVISA Resource Manager with visalib: {self.visalib}"
+                f"Opening PyVISA Resource Manager with visalib: {visalib}"
             )
-            resource_manager = visa.ResourceManager(self.visalib)
-            self.visabackend = self.visalib.split('@')[1]
+            resource_manager = pyvisa.ResourceManager(visalib)
+            visabackend = visalib.split("@")[1]
         else:
             self.visa_log.info("Opening PyVISA Resource Manager with default backend.")
-            resource_manager = visa.ResourceManager()
-            self.visabackend = "ivi"
+            resource_manager = pyvisa.ResourceManager()
+            visabackend = "ivi"
 
-        self.visa_log.info(f'Opening PyVISA resource at address: {address}')
+        self.visa_log.info(f"Opening PyVISA resource at address: {address}")
         resource = resource_manager.open_resource(address)
-        if not isinstance(resource, visa.resources.MessageBasedResource):
-            raise TypeError("QCoDeS only support MessageBasedResource "
-                            "Visa resources")
+        if not isinstance(resource, pyvisa.resources.MessageBasedResource):
+            raise TypeError("QCoDeS only support MessageBasedResource Visa resources")
```

Comment:
I guess this would now leak the resource if it’s not messagebased

Reply:
likely. either leave as is since this hasn't been a problem, or do `resource.close()` ?